ci/env/install-dependencies.sh

```diff
@@ -297,153 +297,163 @@ download_mnist() {
   # dependencies with Modin.
   if [ "${INSTALL_LUDWIG-}" = 1 ]; then
     # TODO: eventually pin this to master.
-    pip install -U "ludwig[test]>=0.4" "jsonschema>=4"
+    requirements_packages+=("ludwig[test]>=0.4")
+    requirements_packages+=("jsonschema>=4")
   fi
 
   # Additional dependency for time series libraries.
   # This cannot be included in requirements_tune.txt as it has conflicting
   # dependencies.
   if [ "${INSTALL_TIMESERIES_LIBS-}" = 1 ]; then
-    pip install -U "statsforecast==1.5.0" "prophet==1.1.1"
+    requirements_packages+=("statsforecast==1.5.0")
+    requirements_packages+=("prophet==1.1.1")
   fi
 
   # Data processing test dependencies.
   if [ "${DATA_PROCESSING_TESTING-}" = 1 ] || [ "${DOC_TESTING-}" = 1 ]; then
-    pip install -U -c "${WORKSPACE_DIR}"/python/requirements.txt -r "${WORKSPACE_DIR}"/python/requirements/data_processing/requirements.txt
+    requirements_files+=("${WORKSPACE_DIR}/python/requirements/data_processing/requirements.txt")
   fi
   if [ "${DATA_PROCESSING_TESTING-}" = 1 ]; then
-    pip install -U -c "${WORKSPACE_DIR}"/python/requirements.txt -r "${WORKSPACE_DIR}"/python/requirements/data_processing/requirements_dataset.txt
+    requirements_files+=("${WORKSPACE_DIR}/python/requirements/data_processing/requirements_dataset.txt")
     if [ -n "${ARROW_VERSION-}" ]; then
       if [ "${ARROW_VERSION-}" = nightly ]; then
-        pip install --extra-index-url https://pypi.fury.io/arrow-nightlies/ --prefer-binary --pre pyarrow
+        delayed_packages+=("--extra-index-url")
+        delayed_packages+=("https://pypi.fury.io/arrow-nightlies/")
+        delayed_packages+=("--prefer-binary")
+        delayed_packages+=("--pre")
+        delayed_packages+=("pyarrow")
       else
-        pip install -U pyarrow=="${ARROW_VERSION}"
+        delayed_packages+=("pyarrow==${ARROW_VERSION}")
       fi
     fi
     if [ -n "${ARROW_MONGO_VERSION-}" ]; then
-	pip install -U pymongoarrow=="${ARROW_MONGO_VERSION}"
+      delayed_packages+=("pymongoarrow==${ARROW_MONGO_VERSION}")
     fi
   fi
 
-  # Remove this entire section once Serve dependencies are fixed.
-  if [ "${MINIMAL_INSTALL-}" != 1 ] && [ "${DOC_TESTING-}" != 1 ] && [ "${TRAIN_TESTING-}" != 1 ] && [ "${TUNE_TESTING-}" != 1 ] && [ "${RLLIB_TESTING-}" != 1 ]; then
-    # If CI has deemed that a different version of Torch
-    # should be installed, then upgrade/downgrade to that specific version.
-    if [ -n "${TORCH_VERSION-}" ]; then
-      case "${TORCH_VERSION-1.9.0}" in
-        1.9.0) TORCHVISION_VERSION=0.10.0;;
-        1.8.1) TORCHVISION_VERSION=0.9.1;;
-        1.5) TORCHVISION_VERSION=0.6.0;;
-        *) TORCHVISION_VERSION=0.5.0;;
-      esac
-      pip install --use-deprecated=legacy-resolver --upgrade torch=="${TORCH_VERSION-1.9.0}" torchvision=="${TORCHVISION_VERSION}"
-    fi
+  if [ "${install_ml_no_deps}" = 1 ]; then
+    pip install -r "${WORKSPACE_DIR}/python/requirements/ml/requirements_no_deps.txt"
```

Comment:
need to specify `--no-deps` here?

Reply:
So the way we do it here is we install _with_ deps, but those deps may be overwritten by the main `pip install` call later. In that sense we don't ensure the dependencies are met after the script ran through. Does that make sense? I've added a comment